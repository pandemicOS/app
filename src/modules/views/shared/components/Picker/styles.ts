import { StyleSheet } from 'react-native';

import { STYLE } from '../../../../../constants';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    paddingHorizontal: 0,
    flexDirection: 'row',
    justifyContent: 'space-between',
    height: 34,
    borderBottomColor: STYLE.COLOR.LIGHTEST_GRAY,
    borderBottomWidth: 1,
    marginTop: 10,
    backgroundColor: STYLE.COLOR.INVISIBLE
  },
  textError: {
    textAlign: 'left',
    color: STYLE.COLOR.RED_CARDINAL,
    backgroundColor: STYLE.COLOR.INVISIBLE
  },
  errorWrapper: {
    marginTop: 8,
    marginBottom: 10
  },
  invalidValue: {
    borderBottomColor: STYLE.COLOR.DARK_RED,
  },
  validValue: {
    borderBottomColor: STYLE.COLOR.LIGHTEST_GRAY,
  },
  text: {
    color: STYLE.COLOR.BLACK,
    fontSize: 15,
    fontWeight: STYLE.FONT.WEIGHTS.MEDIUM,
  }
});

export default styles;
