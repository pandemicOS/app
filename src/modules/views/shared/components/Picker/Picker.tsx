import React from 'react';
import { TouchableOpacity, Text, Image, StyleProp, ViewStyle, View } from 'react-native';

import styles from './styles';

export interface IPickerProps {
  labels: string[];
  values: any[];
  selectedValue: any;
  prompt?: string;
  isDisabled?: boolean;
  onPress: () => void;
  onValidateShowError?: (selectedValue: any) => string;
  onValueChange?: (value: string) => void;
  style?: StyleProp<ViewStyle>;
}

export interface IPickerState {
  isValid: boolean;
  hasValidated: boolean;
  error: string;
}

export default class Picker extends React.Component<IPickerProps, IPickerState> {
  public state = {
    hasValidated: false,
    error: null
  } as IPickerState;

  public isValid = (): boolean => {
    const { onValidateShowError } = this.props;
    const { error, hasValidated } = this.state;
    return onValidateShowError === undefined || (!error && hasValidated);
  }

  public validate = (value: any) => {
    const { onValidateShowError } = this.props;
    if (onValidateShowError) {
      this.setState({ error: onValidateShowError(value), hasValidated: true });
    }
  }

  public resetValidation = () => {
    this.setState({ error: null, hasValidated: false });
  }

  public handlePress = () => {
    this.props.onPress();
  }

  public render() {
    const { selectedValue, style, labels, values, isDisabled, onPress } = this.props;
    const { error } = this.state;
    return (
      <View>
        <TouchableOpacity
          onPress={() => onPress()}
          style={[styles.container, style, error ? styles.invalidValue : styles.validValue, { opacity: isDisabled ? 0.5 : 1 }]}
        >
          <Text style={styles.text}>{selectedValue && labels[values.indexOf(selectedValue)]}</Text>
          <Image
            style={{ width: 15, height: 8, marginRight: 5 }}
            source={require('../../../../../assets/images/down-arrow.png')}
          />
        </TouchableOpacity>
        {error ? (
          <View style={styles.errorWrapper}>
            <Text style={styles.textError}>
              {error}
            </Text>
          </View>
        ) : null}
      </View>
    );
  }
}
