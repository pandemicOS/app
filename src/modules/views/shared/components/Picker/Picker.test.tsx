import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { ActionSheet } from 'native-base';

import Picker, { IPickerProps } from './Picker';
import { View } from 'react-native';

describe('Picker', () => {
  let props: IPickerProps;
  beforeEach(() => {
    props = {
      labels: ['label1', 'label2'],
      values: ['value1', 'value2'],
      selectedValue: '',
      prompt: 'this is a prompt',
      onValueChange: jest.fn(),
      style: {}
    };

    Picker.ActionSheet = { show: jest.fn().mockReturnValue({ onSelect: jest.fn() }) } as any;
  });

  it('renders without crashing', () => {
    const rendered = renderer.create(<Picker {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('renders with error', () => {
    const wrapper = shallow(<Picker {...props} />);
    const elementCount = wrapper.find(View).length;
    wrapper.instance().setState({ error: 'error' });
    expect(wrapper.find(View).length).toBeGreaterThan(elementCount);
  });

  it('should handle onPress', () => {
    const wrapper = shallow(<Picker {...props} />);
    (wrapper.find('TouchableOpacity').first().prop('onPress') as any)();
    expect(Picker.ActionSheet.show).toBeCalled();
  });

  it('should handle onSelect', () => {
    const wrapper = shallow<Picker>(<Picker {...props} />);
    wrapper.instance().onSelect(0);
    expect(props.onValueChange).toBeCalledWith('value1');
    wrapper.instance().onSelect(2);
    expect(props.onValueChange).toHaveBeenCalledTimes(1);
  });

  it('should handle onSelect and validate', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Picker>(<Picker {...props} />);
    wrapper.instance().onSelect(0);
    expect(props.onValueChange).toBeCalledWith('value1');
    wrapper.instance().onSelect(2);
    expect(props.onValueChange).toHaveBeenCalledTimes(1);
  });

  it('should resetValidation', () => {
    const wrapper = shallow<Picker>(<Picker {...props} />);
    wrapper.instance().setState({ error: 'error', hasValidated: true });
    wrapper.instance().resetValidation();
    const newState = wrapper.instance().state;
    expect(newState).toEqual({ error: null, hasValidated: false });
  });

  it('isValid with undefined onValidateShowError should return true', () => {
    const wrapper = shallow<Picker>(<Picker {...props} />);
    expect(wrapper.instance().isValid()).toBe(true);
  });

  it('isValid with error should return false', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Picker>(<Picker {...props} />);
    wrapper.instance().setState({ error: 'error', hasValidated: true });
    expect(wrapper.instance().isValid()).toBe(false);
  });

  it('isValid without error and previously validated should return true', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Picker>(<Picker {...props} />);
    wrapper.instance().setState({ error: null, hasValidated: true });
    expect(wrapper.instance().isValid()).toBe(true);
  });

  it('isValid without error but not previously validated should return false', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Picker>(<Picker {...props} />);
    wrapper.instance().setState({ error: null, hasValidated: false });
    expect(wrapper.instance().isValid()).toBe(false);
  });
});
