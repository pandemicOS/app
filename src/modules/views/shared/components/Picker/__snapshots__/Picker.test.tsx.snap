// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker renders without crashing 1`] = `
<View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "borderBottomColor": "#E7E7E7",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "height": 42,
        "justifyContent": "space-between",
        "marginTop": 10,
        "opacity": 1,
        "paddingHorizontal": 5,
      }
    }
  >
    <Text>
      
    </Text>
    <Image
      source={1}
      style={
        Object {
          "height": 8,
          "marginRight": 5,
          "width": 15,
        }
      }
    />
  </View>
</View>
`;
